### PR TITLE
Creation of Preliminary Bayesian Optimization Algorithm and NuScale Lookup Interface

### PR DESCRIPTION
**Pull request by cahowar9**
Created a Bayesian Optimization algorithm and sample case to run. Also implemented a code interface for the NuScale lookup database stored in hdf5 files on the server. The sample case uses the NuScale lookup tables.

### CODE REVIEW

File: midas/input_parser.py
Comment (nick-rollins-ncsu) [inline midas/input_parser.py:59]: "self" is not passed to validate_input(), also, self.methodology has not been defined yet. The value you're wanting to used is stored in "value".

File: midas/input_parser.py
Comment (nick-rollins-ncsu) [inline midas/input_parser.py:495]: what is batch_size used for? how does it differ from population_size?

File: midas/codes/nuscale_lut.py
Comment (nick-rollins-ncsu) [inline midas/codes/nuscale_lut.py:51]: where do you require that the user provide these input options? what if a different option is provided? I think you might need additional logic in the input_parser.py to enforce this.

File: midas/input_parser.py
Comment (nick-rollins-ncsu) [inline midas/input_parser.py:59]: "self" is not passed to validate_input(), also, self.methodology has not been defined yet. The value you're wanting to used is stored in "value".

File: midas/input_parser.py
Comment (nick-rollins-ncsu) [inline midas/input_parser.py:495]: what is batch_size used for? how does it differ from population_size?

**Review by nick-rollins-ncsu — CHANGES_REQUESTED** [review]
Bayesian opt looks good. I think the Nuscale LUT needs more rigorous testing to ensure invalid inputs are correctly resolved. Please correct me if this has already been accounted for.

File: midas/input_parser.py
Comment (cahowar9) [reply midas/input_parser.py:495]: batch size is essentially the same as population size, where it represents the number of individuals created and evaluated per generation. However, I intend to use this in order to have a population size number of initial random points and then batch size number of individuals per generation. The reason is because the number of initial points and number of points per iteration can affect the optimization, where sometimes having a relatively small or large batch size can help it focus more on exploring a wider range of solutions or in finding a more precise solution.

File: midas/codes/nuscale_lut.py
Comment (cahowar9) [reply midas/codes/nuscale_lut.py]: I just pushed a new version of this into my fork, I believe you should see it in the current pull request but I can create a new one if need be. I changed the logic so the fuel assembly name has no bearing on the optimization itself, and takes the "type" parameter under fuel in the assembly options for the corresponding fuel assembly name. I also added a check to make sure it's a valid type (i.e. between 2-7), but can move this logic to input_parser or somewhere else if need be. Let me know if anything else here needs to be changed

Comment (cahowar9) [reply midas/codes/nuscale_lut.py]: I can certainly change this if you'd like me to. However, this should work with any conceivable name a user might input for the fuel assembly name, with my most recent push. It would print the chromosomes as the user-defined names (in this case something like NSFA455), and this code interface would look into the genome/fa options and match it with the corresponding fuel type (2-7) regardless of how the user defined it. I believe validate_input already checks to make sure that each fuel assembly in decision variables has a matching fuel parameter under fa options, so this is likely to work as is. A quick question, I saw that my changes were approved, is there something I need to do on my end to merge the branches, or are you waiting for me to address these additional comments before merging them?

File: midas/algorithms/bayesian_optimization.py
Comment (cahowar9) [reply midas/algorithms/bayesian_optimization.py:19]: Understood. I plan to go back through my code at some point and thoroughly comment everything. The reason I have a parse_dimensions function is because the bayesian optimization needs to know what values each individual and chromosome can take. This library specifically needs it in a specific format, where each chromosome has a list of what values it can take, preceded by the value type (integer, real, or categorical). This function reads the genome and formats the possible values in this way for the optimizer object to understand. I hope this helps.

File: midas/codes/nuscale_lut.py
Comment (nick-rollins-ncsu) [reply midas/codes/nuscale_lut.py]: I approved the pull request because I didn't think my additional comments were vital, but I didn't merge the request in case you wanted to make any final changes based on my comments. Once approval has been given, you should be able to initiate the merge at your convenience.

Comment (cahowar9) [reply midas/codes/nuscale_lut.py]: Understood. I'd like to go ahead and merge it now and address some of these comments later, so that others can use the database if they'd like. However, I do not have write access, so even with the approved pull request I believe I am unable to merge these myself. Would it be possible for you to merge these branches now and then I can come back to these comments later?

**Pull request merged**
